kernel/sched.c

```diff
@@ -788,7 +788,9 @@ static inline void unpend_thread_no_timeout(struct k_thread *thread)
 ALWAYS_INLINE void z_unpend_thread_no_timeout(struct k_thread *thread)
 {
 	LOCKED(&sched_spinlock) {
-		unpend_thread_no_timeout(thread);
+		if (thread->base.pended_on != NULL) {
```

Comment:
Not understanding the logic change here.  The current (not terribly informative) naming scheme is that z_xxxx() scheduler APIs are (or at least tend to be) synchronized calls and the internal static functions have no prefix and (usually) are expected to be externally synchronized.

And that's true here, where this used to be just a wrapper around the internal function.  But now it's got different behavior, which seems likely to be confusing.

What's the circumstances where you are now able to land here to unpend a thread that isn't actually pended?  That seems like a state management problem somewhere else?

Reply:
Excellent question--and it is for an SMP case that I did not originally spot when I first re-did the pipes implementation.

Pipe operations essentially act in two stages. In stage one, we create a list of the copying work that needs to be done. In stage two, we perform the copy and unpend an item if necessary once the copying is complete. However, unless I am misunderstanding something, there exists a window of opportunity where a timeout can be processed on another core after the work list has been generated that can cause it to be removed from the wait/pend queue. The pipe's spin-lock does not protect against this.